PARCEL_2_RFC.md

```diff
@@ -0,0 +1,1285 @@
+
+```
+/node_modules/.cache/parcel/
+  /00/
+    213debd8ddd45819b79a3a974ed487.json
+    40ae9b581afc53841307a4b3c2463d.json
+    63a9dd58fc1e8f8bb819759ea9793c.json
+    ...
+  /01/
+  /../
+  /zy/
+  /zz/
+```
+
+It follows this weird structure in order to avoid too many files being created
+in a single directory which degrades file system performance.
+
+## Asset Resolution
+
+Parcel follows the Node module resolution algorithm with a few additions.
+
+### Local Paths
+
+```
+./path/to/file
+./path/to/file.js
+```
+
+These follow the [Node module resolution algorithm](https://nodejs.org/api/modules.html#modules_all_together).
+
+### Package Paths
+
+```
+preact
+lodash/cloneDeep
+@sindresorhus/is
+```
+
+These follow the [Node module resolution algorithm](https://nodejs.org/api/modules.html#modules_all_together).
+
+### URLs
+
+```
+https://unpkg.com/preact@8.2.9/dist/preact.min.js
+```
+
+Parcel by default will ignore URL dependencies, other resolver plugins may
+choose to do something with them.
+
+### Tilde Paths
+
+```
+~/src/file.js
+```
+
+Only when used outside of `node_modules` directories, the `~` is replaced by an
+absolute path to the closest package root:
+
+```sh
+/path/to/app #(/package.json)
+```
+
+To form a path that looks like:
+
+```
+/path/to/app/src/file.js
+```
+
+Then it follows the [Node module resolution algorithm](https://nodejs.org/api/modules.html#modules_all_together).
+
+### Aliases
+
+Aliases come in two forms:
+
+1. Package Aliases: `react -> preact`
+2. File/Directory Aliases: `utils` -> `./src/utils`
+
+```json
+{
+  "name": "my-project",
+  "alias": {
+    "react": "preact-compat",
+    "react-dom": "preact-compat",
+    "utils": "./src/utils",
+    "components": "./src/components"
+  }
+}
+```
+
+There are a couple of rules:
+
+1. Aliases will only be respected when specified outside of `node_modules`.
+2. Aliases specified outside of `node_modules` will affect assets inside of `node_modules`.
+3. Aliases cannot build off of other aliases.
+4. Only one alias will be applied at a time.
+5. Aliases must be valid npm package names.
+
+## Plugins
+
+### Resolvers
+
+When one asset depends on another through an asset specifier, the resolver is
+responsible for determining what asset is being requested.
+
+See [Asset Resolution](#asset-resolution) for more details.
+
+```json
+{
+  "resolvers": ["@parcel/resolver-v1"]
+}
+```
+
+**Official Resolvers:**
+
+- `@parcel/resolver-v1`
+
+### Transforms
+
+Transforms _transform_ single assets as they are discovered and added to the
+asset graph. They mostly call out to different compilers and preprocessors.
+
+```json
+{
+  "transforms": {
+    "*.js": ["@parcel/transform-babel"]
+  }
+}
+```
+
+**Official Transforms:**
+
+- `@parcel/transform-babel`
+- `@parcel/transform-coffeescript`
+- `@parcel/transform-graphql`
+- `@parcel/transform-json`
+- `@parcel/transform-json5`
+- `@parcel/transform-less`
+- `@parcel/transform-posthtml`
```

Comment:
how can one extend/configure the official transformers, or make use in their own plugins? for example, I want to rewrite some HTML tags with the _posthtml_ transformer?

Reply:
Parcel takes in an array of transforms, so they're chained so you can manipulate the output (of the previous transform) as much as you want.

ASTs are being sent from transform to transform, so if it's compatible you can re-use it (for performance).

Not sure why you'd ever wanna extend a transform, you could possibly re-use internals of another transform although that is probably risky as that's usually outside the scope of the transforms public interface and therefore not protected from change in minor releases.

Reply:
Take the example, where I want to rewrite some tags in the html files. I assume parcel will use the ootb posthtml transformer for finding the referenced assets (CSS,JS). so I would just add my own transformer, based on posthtml, or traverse the AST myself? And add this to the transformer chain for HTML via API or config?

Reply:
I have no clue what ootb means, but this is how it would probably work.

.parcelrc
```json
{
  "extends": "@parcel/config-default",
  "transforms": {
  	"*.html": ["@parcel/transform-posthtml", "parcel-transform-something"]
  }
}
```

Parcel will run through it's own stuff collecting deps,... than `parcel-transform-something` would run in the same thread on the same asset, with the output of the previous transform as it's input. You could traverse the ast to get some custom dependencies and I think you can also modify dependencies.

This is how plugins work posthtml as well, so it should be no different. Everything should be possible.

Reply:
The API will also search for the parcelrc file or default config as far as I know.

It will also take an options object similar to the one parcel uses now but slightly different as you can see some cli flags are different in this RFC.

Not sure if you'll be able to provide a hardcoded config.

Reply:
> Not sure if you'll be able to provide a hardcoded parcelrc config.

We want to use parcel as packager within our CLI tool, and ensure that certain transformers are present. with the 1.x version, we just register the asset types directly via API. so it would be important to be able to do the same for the transformers as well. 

something along the lines of:

```
let tx = bundler.getTransformers('*.html');
tx.push(myCoolTransformer);
bundler.setTransformers('*.html');
```

(but maybe we should move this discussion into an issue)

Reply:
No need to create a new issue, this is one big RFC for parcel 2.
So it's already in the right place.

The API should probably also be documented in this spec.